base_layer/core/src/base_node/state_machine_service/handle.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::base_node::state_machine_service::states::{StateEvent, StatusInfo};
+use core::sync::atomic::{AtomicBool, Ordering};
+use futures::{stream::Fuse, StreamExt};
+use std::sync::Arc;
+use tari_broadcast_channel::Subscriber;
+use tari_shutdown::ShutdownSignal;
+use tokio::sync::watch;
+
+#[derive(Clone)]
+pub struct StateMachineHandle {
+    state_change_event_subscriber: Subscriber<StateEvent>,
+    status_event_receiver: watch::Receiver<StatusInfo>,
+    shutdown_signal: ShutdownSignal,
+    initial_blockchain_sync_success: Arc<AtomicBool>,
+}
+
+impl StateMachineHandle {
+    pub fn new(
+        state_change_event_subscriber: Subscriber<StateEvent>,
+        status_event_receiver: watch::Receiver<StatusInfo>,
+        initial_blockchain_sync_success: Arc<AtomicBool>,
```

Comment:
The `AtomicBool` seems out of place here. Is it not better to use channels similar to the rest of the struct?

Reply:
I disagree on this one. This is an event that occurs exactly once and is not a stream of changes so I believe a read-only AtomicBool is appropriate.

Reply:
Agree with @philipr-za